### ISSUE TITLE
Unhandled Exception in `connection.end()` for MySQL Dialect

### ISSUE THREAD

**Issue by brycekbargar** [tgriesser/knex#1691]
[There's an open issue on the mysql repo](https://github.com/mysqljs/mysql/issues/1334) for ClearDB not responding the a Quit packet.

This affects knex when destroying an idle connection using [`Client_MySQL.destroyRawConnection()`](https://github.com/tgriesser/knex/blob/master/src/dialects/mysql/index.js#L80).

```js
  destroyRawConnection(connection) {
    // This removes the error handler attached by knex in `aquireRawConnection()`
    // It also removes any error handlers attached by the end user
    connection.removeAllListeners()
    // In the mysql library this takes an error callback, or alternatively emits the 'error' event
    connection.end()
  }
```

The unhandled exception thrown by mysql on `connection.end()` was crashing my application and there is really no way to handle it as knex is currently setup.

I would be willing to open a PR and fix the bug, but I'm not sure which strategy is most appropriate.

The easiest would be to remove the call to `removeAllListeners()`, [according to the NodeJS docs](https://nodejs.org/api/events.html#events_emitter_removealllisteners_eventname)

> "Note that it is bad practice to remove listeners added elsewhere in the code, particularly when the EventEmitter instance was created by some other component or module (e.g. sockets or file streams)."

That said, I'm assuming there is a reason knex is calling this method? Maybe something to do with performance?

An alternative would be to pass in an error callback to `.end()', for example:
```js
  destroyRawConnection(connection) {
    connection.removeAllListeners()
    connection.end(err => {
      connection.__knex__disposed = err
    })
  }
```

Let me know which approach would be best and I'll make the change and open a PR.

As a workaround, for anyone else experiencing issues with knex, heroku, mysql, and cleardb I've done the following and it works for now but is fairly ugly...

```js
config.pool.afterCreate = (conn, cb) => {
  const removeAllListeners = conn.removeAllListeners;
  conn.removeAllListeners = () => {
    removeAllListeners.call(conn);
    conn.on('error', err => conn.__knex__disposed = err);
  };
  cb(null);
};
```

**Comment by tgriesser** [comment]
> The unhandled exception thrown by mysql on connection.end() was crashing my application and there is really no way to handle it as knex is currently setup.

Yikes, sorry about that.

> That said, I'm assuming there is a reason knex is calling this method? Maybe something to do with performance?

Basically it's there just to clean up any listeners hanging around that aren't needed anymore since the connection shouldn't be accessible for any purpose once it's "destroyed" anyway.

I think this would be the preferred solution:

```js
connection.end(err => {
  err.__knex__disposed = err
})
``` 

or maybe even:

```js
destroyRawConnection(connection) {
  connection.end(() => {
    connection.removeAllListeners()
  })
}
```

so that any error handler still exists until after the connection is definitely ended.

**Comment by tgriesser** [comment]
Just pushed a fix in 0.12.2, thanks for the report!
